mozetl/taar/taar_locale.py

```diff
@@ -0,0 +1,177 @@
+    key_path = s3_prefix + s3_dest_file_name
+    transfer.upload_file(source_file_name, bucket, key_path)
+
+
+def get_addons(spark):
+    """ Longitudinal sample is selected and freshest ping chosen per client.
+    Only Firefox release clients are considered.
+    Columns are exploded (over addon keys)  to include locale of each addon
+    installation instance system addons, disabled addons, unsigned addons
+    are filtered out.
+    Sorting by addon-installations and grouped by locale.
+    """
+    return spark.sql("""
+        WITH sample AS (
+        SELECT client_id,
+        settings[0].locale AS locality,
+        EXPLODE(active_addons[0])
+        FROM longitudinal
+        WHERE normalized_channel='release'
+          AND build IS NOT NULL
+          AND build[0].application_name='Firefox'
+        ),
+
+        filtered_sample AS (
+        SELECT locality, key AS addon_key FROM sample
+        WHERE value['blocklisted'] = FALSE -- not blocklisted
+          AND value['type'] = 'extension' -- nice webextensions only
+          AND value['signed_state'] = 2 -- fully reviewed addons only
+          AND value['user_disabled'] = FALSE -- active addons only get counted
+          AND value['app_disabled'] = FALSE -- exclude compatibility disabled addons
+          AND value['is_system'] = FALSE -- exclude system addons
+          AND locality <> 'null'
+          AND key is not null
+        ),
+
+        country_addon_pairs AS (
+        SELECT
+        COUNT(*) AS pair_cnts, addon_key, locality
+        from filtered_sample
+        GROUP BY locality, addon_key
+        )
+
+        SELECT * FROM country_addon_pairs
+        ORDER BY locality, pair_cnts DESC
+    """)
+
+
+def compute_threshold(addon_df):
+    """ Get a threshold to remove locales with a small
+    number of addons installations.
+    """
+    addon_install_counts = addon_df.groupBy('locality')\
+        .agg({'pair_cnts': 'sum'})\
+        .rdd.map(lambda p: p[1])\
+        .collect()
+
+    # Compute a threshold at the 25th percentile to remove locales with a
+    # small number of addons installations.
+    locale_pop_threshold = np.percentile(np.array(addon_install_counts), 25)
+
+    # Safety net in case the distribution gets really skewed, we should
+    # require 2000 addon installation instances to make recommendations.
+    return 2000 if locale_pop_threshold < 2000 else locale_pop_threshold
+
+
+def transform(addon_df, threshold, num_addons):
+    """ Converts the locale-specific addon data in to a dictionary.
+
+    :param addon_df: the locale-specific addon dataframe;
+    :param threshold: the minimum number of addon-installs per locale;
+    :param num_addons: requested number of recommendations.
+    :return: a dictionary {<locale>: ['GUID1', 'GUID2', ...]}
+    """
+    top10_per = {}
+
+    # Decide that we can not make reasonable recommendations without
+    # a minimum number of addon installations.
+    grouped_addons = addon_df\
```

Comment:
Style nit: prefer implicit line continuations in parenthesis over backslashes, as per [pep8](https://www.python.org/dev/peps/pep-0008/#maximum-line-length). Here, it would look like this:

```
grouped_addons = (
    addon_df
    .groupBy('locality')
    .agg({'pair_cnts': 'sum'})
    .collect()
)

Reply:
Done!